pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,278 @@
+			return fmt.Errorf("failed to get Azure Cluster Identity: %w", err)
+		}
+
+		azureClientID, ok := capzManagerBootstrapSecret.Data["azure_client_id"]
+		if !ok {
+			return fmt.Errorf("failed to get azureClientID: %w", err)
+		}
+
+		azureTenantID, ok := capzManagerBootstrapSecret.Data["azure_tenant_id"]
+		if !ok {
+			return fmt.Errorf("failed to get azureTenantID: %w", err)
+		}
+
+		azureClusterIdentity = &azurev1.AzureClusterIdentity{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      r.Infra.Status.InfrastructureName,
+				Namespace: defaultCAPINamespace,
+				Annotations: map[string]string{
+					// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
+					// as that's managed externally, in this case by the cluster-capi-operator's infracluster controller.
+					clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+				},
+			},
+			Spec: azurev1.AzureClusterIdentitySpec{
+				Type:              azurev1.ServicePrincipal,
+				AllowedNamespaces: &azurev1.AllowedNamespaces{NamespaceList: []string{defaultCAPINamespace}},
+				ClientID:          string(azureClientID),
+				TenantID:          string(azureTenantID),
+				ClientSecret:      corev1.SecretReference{Name: clusterSecretName, Namespace: defaultCAPINamespace},
+			},
+		}
+
+		// The Azure Cluster Identtiy does not exist, so it needs to be created.
+		if err := r.Create(ctx, azureClusterIdentity); err != nil && !cerrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create Azure Cluster Identity: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// ensureAzureInfraCluster ensures the InfraCluster exists, and if it doesn't, creates it.
+//
+//nolint:funlen
+func (r *InfraClusterController) ensureAzureInfraCluster(ctx context.Context, target *azurev1.AzureCluster, log logr.Logger) error {
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	// Checking whether InfraCluster object exists. If it doesn't, create it.
+	if err := r.Get(ctx, client.ObjectKeyFromObject(target), target); err != nil {
+		if !cerrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get InfraCluster: %w", err)
+		}
+
+		log.Info(fmt.Sprintf("AzureCluster %s/%s does not exist, creating it", target.Namespace, target.Name))
+
+		providerSpec, err := getAzureMAPIProviderSpec(ctx, r.Client)
+		if err != nil {
+			return fmt.Errorf("error obtaining Azure Provider Spec: %w", err)
+		}
+
+		location, err := r.getAzureLocation(ctx)
+		if err != nil {
+			return fmt.Errorf("error obtaining Azure Cluster location: %w", err)
+		}
+
+		target = &azurev1.AzureCluster{
```

Comment:
Separating the creation of this azure cluster object (I mean literally instantiating the object and all the fields within) into a separate function would make this cleaner